src/lib/apply-patch/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+                        .0
+                        + 1
+                };
+                // start..end should be the byte range in `content` to be patched
+                let end = lines
+                    .nth(expected_lines.len() - 1)
+                    .ok_or(MismatchError::NotEnoughLines)?
+                    .0
+                    + 1;
+                if compare_range(
+                    &content[start..end],
+                    &expected_lines,
+                    self.ignore_whitespace,
+                ) {
+                    start..end
+                } else {
+                    return Err(MismatchError::LineMismatch);
+                }
+            };
+            // first copy to the output all the content between either the last patched range or the beginning of the file, and the beginning of this patch
+            new_content.push_str(&content[prev_idx..found_range.start]);
+            prev_idx = found_range.end;
+            // then interleave the context lines with the added lines
+            let mut orig_lines = content[found_range].lines();
+            for hunk_line in &hunk.lines {
+                match *hunk_line {
+                    Line::Context(_) => {
+                        // in the case of a context line, push the original line, not the one from the patch file
+                        // this may be a whitespace-insensitive patch, and we don't want to modify any lines that aren't marked `-`
+                        new_content.push_str(orig_lines.next().unwrap());
+                        new_content.push('\n');
+                    }
+                    Line::Add(s) => {
+                        new_content.push_str(s);
+                        new_content.push('\n');
+                    }
+                    Line::Remove(_) => {
+                        orig_lines.next().unwrap();
+                    }
+                }
+            }
+        }
+        // finally, copy everything between the final patch and the end of the file
+        new_content.push_str(&content[prev_idx..]);
+        Ok(new_content)
+    }
+}
+
+fn compare_range(content: &str, lines: &[&str], ignore_whitespace: bool) -> bool {
+    for (left, &right) in content.lines().zip(lines) {
+        if ignore_whitespace {
+            if left != right {
+                return false;
+            }
+        } else if !left
+            .chars()
+            .filter(|ch| !ch.is_whitespace())
+            .eq(right.chars().filter(|ch| !ch.is_whitespace()))
+        {
+            return false;
+        }
+    }
+    true
+}
+
+impl Default for Settings {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum MismatchError {
+    #[error("File did not contain enough lines")]
+    NotEnoughLines,
+    #[error("Mismatch between context/removal line and file")]
+    LineMismatch,
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error(transparent)]
+    Io(#[from] std::io::Error),
```

Comment:
`std::io::Error` generally doesn't have enough contextual information to be useful by itself.  I would only add std::io::Error as a part of another type that does have context.
```suggestion
```

Reply:
I'll just delete this then, it's a placeholder for nonexistent-as-of-yet functionality anyway.